test/gtest/ucp/test_ucp_request.cc

```diff
@@ -100,3 +102,392 @@ UCS_TEST_P(test_ucp_request, test_request_query)
 }
 
 UCP_INSTANTIATE_TEST_CASE_TLS(test_ucp_request, all, "all")
+
+class test_proto_reset : public ucp_test {
+public:
+    enum operation_e {
+        TAG,
+        RMA_GET,
+        RMA_PUT,
+        STREAM,
+        AM
+    };
+
+    test_proto_reset() : m_msg_size(UCS_KBYTE * 64), m_completed(0)
+    {
+        m_pending.head  = NULL;
+        m_pending.ptail = NULL;
+    }
+
+    void init() override
+    {
+        if (!m_ucp_config->ctx.proto_enable) {
+            UCS_TEST_SKIP_R("reset is not supported for proto v1");
+        }
+
+        ucs::scoped_setenv snd_buf("UCX_TCP_SNDBUF", "8K");
+        ucs::scoped_setenv ib_queue_len("UCX_IB_TX_QUEUE_LEN", "65");
+        ucp_test::init();
+        connect();
+    }
+
+    void cleanup() override
+    {
+        for (auto &rkey : m_rkeys) {
+            ucp_rkey_destroy(rkey);
+        }
+
+        for (auto &memh : m_memhs) {
+            ASSERT_UCS_OK(ucp_mem_unmap(receiver().ucph(), memh));
+        }
+
+        ucp_test::cleanup();
+    }
+
+    void connect()
+    {
+        ucp_request_param_t param = {0};
+        param.op_attr_mask        = UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+        static const size_t size  = 1024;
+        std::vector<char> sbuf(size);
+        std::vector<char> rbuf(size);
+
+        sender().connect(&receiver(), get_ep_params());
+        receiver().connect(&sender(), get_ep_params());
+
+        void *pre_msg = ucp_tag_send_nbx(sender().ep(), sbuf.data(), size, 0,
+                                         &param);
+        void *pre_rcv = ucp_tag_recv_nbx(receiver().worker(), rbuf.data(), size,
+                                         0, 0, &param);
+        request_wait(pre_msg);
+        request_wait(pre_rcv);
+    }
+
+    void get_stream_data(std::vector<char> &rbuf)
+    {
+        size_t roffset            = 0;
+        const double timeout      = 10;
+        const ucs_time_t deadline = ucs::get_deadline(timeout);
+        ucs_status_ptr_t rdata;
+        size_t length;
+
+        do {
+            progress();
+            rdata = ucp_stream_recv_data_nb(receiver().ep(), &length);
+            if (rdata == NULL) {
+                continue;
+            }
+
+            memcpy(&rbuf[roffset], rdata, length);
```

Comment:
In general it's not safe to append elements to vector like that, it disrespects vector invariants.
Why not using std::vector::insert that accepts a range?

Reply:
I use a different API now (ucp_stream_recv_nbx) so memcpy is not needed at all